packages/gatsby-dev-cli/src/local-npm-registry/verdaccio.js

```diff
@@ -0,0 +1,171 @@
+          // to local registry.
+          console.log(`Logging in with dummy user`)
+
+          npmLogin(
+            `gatsby-dev`,
+            `gatsby-dev`,
+            `gatsby-dev@gatsbyjs.org`,
+            registryUrl
+          )
+
+          // This is silly - `npm-cli-login` doesn't return promise or callback
+          // so let's just wait reasaonable amount before resolving
+          // There is open pull request to add returning promise:
+          // https://github.com/postmanlabs/npm-cli-login/pull/19
+          setTimeout(() => {
+            console.log(`Hopefully logged in already`)
+            resolve()
+          }, 1500)
+        })
+      }
+    )
+  })
+
+  return VerdaccioInitPromise
+}
+
+exports.startVerdaccio = startServer
+
+const publishPackage = async ({
+  packageName,
+  packagesToPublish,
+  root,
+  versionPostFix,
+}) => {
+  // we need to check if package depend on any other package to will be published and
+  // adjust version selector to point to dev version of package so local registry is used
+  // for dependencies.
+
+  const monoRepoPackageJsonPath = getMonorepoPackageJsonPath({
+    packageName,
+    root,
+  })
+  const monorepoPKGjsonString = fs.readFileSync(monoRepoPackageJsonPath)
+  const monorepoPKGjson = JSON.parse(monorepoPKGjsonString)
+
+  monorepoPKGjson.version = `${monorepoPKGjson.version}-dev-${versionPostFix}`
+  packagesToPublish.forEach(packageThatWillBePublished => {
+    if (monorepoPKGjson.dependencies[packageThatWillBePublished]) {
+      // change to "gatsby-dev" dist tag
+      monorepoPKGjson.dependencies[packageThatWillBePublished] = `gatsby-dev`
+    }
+  })
+
+  // change version and dependency versions
+  fs.outputJSONSync(monoRepoPackageJsonPath, monorepoPKGjson)
+
+  const pathToPackage = path.dirname(monoRepoPackageJsonPath)
+
+  // npm publish
+  const publishCmd = [
+    `npm`,
+    [`publish`, `--tag`, `gatsby-dev`, `--registry=${registryUrl}`],
+    {
+      cwd: pathToPackage,
+    },
+  ]
+
+  console.log(
+    `Publishing ${packageName}@${monorepoPKGjson.version} to local registry`
+  )
+  try {
+    await promisifiedSpawn(publishCmd)
+
+    console.log(
+      `Published ${packageName}@${monorepoPKGjson.version} to local registry`
+    )
+  } catch {
+    console.error(`Failed to publish ${packageName}@${monorepoPKGjson.version}`)
+  }
+
+  // restore original package.json
+  fs.outputFileSync(monoRepoPackageJsonPath, monorepoPKGjsonString)
+}
+
+exports.publishPackagesLocallyAndInstall = async ({
+  packagesToPublish,
+  packages,
+  root,
+}) => {
+  await startServer()
+
+  const versionPostFix = Date.now()
+
+  for (let packageName of packagesToPublish) {
+    await publishPackage({
+      packageName,
+      packagesToPublish,
+      root,
+      versionPostFix,
+    })
+  }
+
+  // TO-DO install packages locally
+  // need to install intersection of `packagesToPublish` and `packages`
+  // (packages filtered out will be deps of packages to install)
```

Comment:
this will cause changes to package.json and yarn.lock in local site (yarn doesn't support `--no-save` - https://github.com/yarnpkg/yarn/issues/1743). I tend to lean toward documenting this behaviour instead of hacking things (storing content of those files and restoring it after installation) as this could cause some weirdness with yarn as it does use those files as source of truth